pkgs/top-level/module/module.nix

```diff
@@ -0,0 +1,174 @@
+/*
+  A [module] for invoking Nixpkgs.
+
+  This module is a public interface for use in third party module system
+  applications, so that they can re-expose the Nixpkgs configuration logic to
+  their users.
+
+  It declares its options without a prefix like `nixpkgs.*`, so it is meant to
+  be imported in a submodule.
+
+  This module is based on the NixOS `misc/nixpkgs.nix` module, but it gets rid
+  of legacy options. It may be used to replace that module after legacy options
+  have been deprecated and removed.
+
+  As this module declares an option named `config`, you're recommended to pass
+  `shorthandOnlyDefinesConfig = true` to `types.submoduleWith`.
+
+  module: https://nixos.org/manual/nixpkgs/unstable/#module-system
+*/
+{ lib, config, options, ... }:
+
+let
+  inherit (lib) mkOption types literalExpression;
+
+  optDefaultPrio = (lib.mkOptionDefault null).priority;
+
+  isCross = config.buildPlatform != config.hostPlatform;
+  systemArgs =
+    if isCross
+    then {
+      localSystem = config.buildPlatform;
+      crossSystem = config.hostPlatform;
+    }
+    else {
+      localSystem = config.hostPlatform;
+    };
+
+  pkgs =
+      import ../default.nix (
+        # NOTE: make inputsSet covers all parameters; see _memoize
+        #       Convenience options that write to inputsSet should be omitted.
+        #       This point is that inputsSet must form a unique key so that
+        #       memoization optimizations are safe.
+        {
+          inherit (config) config overlays;
+        } // systemArgs
+      );
+
+  # TODO: write a better type that checks that the definition ordering is
+  #       unambiguous under import order.
+  overlayType = types.functionTo (types.functionTo types.raw);
+
+  # A unique key for the Nixpkgs package set to be instantiated.
+  #
+  # This is only relevant for _memoize - ie when you're writing or troubleshooting
+  # a module system application that uses _memoize to memoize things.
+  #
+  # Default values and unset values are omitted so that future additions to the
+  # key are usable with _memoize implementations that are not aware of them,
+  # as long as these additions are unused.
+  inputsSet = lib.filterAttrs
+    (_: v: v != v.isDefined && v.highestPrio < optDefaultPrio)
```

Comment:
This condition needs a bit of work, `v != v.isDefined` seems wrong (detected by @roberth himself!)